### PR TITLE
ID token signing algorithm negotiation

### PR DESCRIPTION
**Pull request by NSeydoux**
This enforces negotiation of the algorithm used to sign ID tokens. In its discovery document (`/.well-known/openid-configuration`), the OIDC provider must specify a list of supported signature algorithms with the `id_token_signing_alg_values_supported` property. When registering, a client may specify a signing algorithm that they want used to sign the ID Token the OIDC provider will send. 

The default behaviour is to use `RS256`, but this default behaviour is only a suggestion, not a requirement, which means we cannot rely on the default. Case in point, the ESS ID broker defaults to `ES256`, which breaks the client library if not handled properly. Now, we specify a list of preferred signing algorithms (currently `["ES256", "RS256"]`), and the library selects the most preferred algorithm in the list of algorithms supported by the OIDC issuer.  

Note: the ID token verification is only updated in the `node` library, because the `browser` library doesn't require the signing algorithm to be specified to verify. 

# New feature description

# Checklist

- [X] All acceptance criteria are met.
- [ ] Relevant documentation, if any, has been written/updated.
- [x] The changelog has been updated, if applicable.
- [X] New functions/types have been exported in `index.ts`, if applicable.
- [X] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

### CODE REVIEW

File: packages/core/src/login/oidc/IClientRegistrar.ts [hunk 1]
```diff
@@ -58,6 +58,22 @@ function isValidUrl(url: string): boolean {
...
+  let preferredSupportedAlg = null;
+  preferred.some((signingAlg) => {
+    if (supported.includes(signingAlg)) {
+      preferredSupportedAlg = signingAlg;
+      // returnning true breaks the .some() iteration.
+      return true;
+    }
+    return false;
+  });
+  return preferredSupportedAlg;
```
Comment (ajacksified) [inline packages/core/src/login/oidc/IClientRegistrar.ts:74]: This could be written more simply using `.find`:

```
export function determineSigningAlg(
  supported: string[],
  preferred: string[]
): string | null {
  return preferred.find((signingAlg) => {
    return supported.includes(signingAlg);
  }) || null;
}
```

File: packages/oidc/src/common/negociation.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  let preferredSupportedAlg = null;
+  preferred.some((signingAlg) => {
+    if (supported.includes(signingAlg)) {
+      preferredSupportedAlg = signingAlg;
+      // returnning true breaks the .some() iteration.
+      return true;
+    }
+    return false;
+  });
+  return preferredSupportedAlg;
```
Comment (ajacksified) [inline packages/oidc/src/common/negociation.ts:39]: Same here - logic can be simplified with `find`

**Review by ajacksified — CHANGES_REQUESTED** [review]

File: packages/core/src/login/oidc/IClientRegistrar.ts (hunk 1 above)
Comment (NSeydoux) [reply packages/core/src/login/oidc/IClientRegistrar.ts]: d7febeaa8dc187ada7ce826058fd42a4c1dbe7b5

File: packages/oidc/src/common/negociation.ts (hunk 2 above)
Comment (NSeydoux) [reply packages/oidc/src/common/negociation.ts]: 498eeaad6f60d848cf6118e3103b531e89f1cc25

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,19 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
 
 ## Unreleased
 
+### Bugfixes
+
+#### node and oidc
+
+- The OIDC issuer profile is used to negociate the preferred signature algorithm
```
Comment (pmcb55) [inline CHANGELOG.md:14]:
```suggestion
- The OIDC issuer profile is used to negotiate the preferred signature algorithm
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,6 +7,19 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
...
+for ID tokens.
+
+#### node
+
+- During client registration, the client explicitly registers both the 'refresh_token'
```
Comment (Vinnl) [inline CHANGELOG.md:19]: Minor, but as a developer who just wants to be able to connect to people's Pods, I have no idea how this affects me :P How do I know if this bug affected me, and if it did, how?

File: packages/core/src/login/oidc/IClientRegistrar.ts [hunk 5]
```diff
@@ -58,6 +58,17 @@ function isValidUrl(url: string): boolean {
...
 }
 
+export function determineSigningAlg(
+  supported: string[],
+  preferred: string[]
+): string | null {
+  return (
+    preferred.find((signingAlg) => {
+      return supported.includes(signingAlg);
+    }) || null
```
Comment (Vinnl) [inline packages/core/src/login/oidc/IClientRegistrar.ts:68]: Theoretically shouldn't cause any problems here, but generally good practice to avoid casting if unnecessary.

```suggestion
    }) ?? null
```

File: packages/core/src/login/oidc/IClientRegistrar.test.ts [hunk 6]
```diff
@@ -109,3 +109,19 @@ describe("handleRegistration", () => {
...
+    expect(
+      determineSigningAlg(["ES256", "HS256", "RS256"], ["ES256", "RS256"])
+    ).toBe("ES256");
+    expect(determineSigningAlg(["ES256", "HS256", "RS256"], ["RS256"])).toBe(
+      "RS256"
+    );
+    expect(determineSigningAlg(["RS256"], ["RS256"])).toBe("RS256");
+  });
+
+  it("returns undefined if there are no matches", () => {
```
Comment (Vinnl) [inline packages/core/src/login/oidc/IClientRegistrar.test.ts:124]:
```suggestion
  it("returns null if there are no matches", () => {
```

File: packages/node/src/login/oidc/ClientRegistrar.ts [hunk 7]
```diff
@@ -32,10 +32,40 @@ import {
...
+  PREFERRED_SIGNING_ALG,
 } from "@inrupt/solid-client-authn-core";
 import { Client, Issuer } from "openid-client";
 import { configToIssuerMetadata } from "./IssuerConfigFetcher";
 
+export function negotiateClientSigningAlg(
+  issuerConfig: IIssuerConfig,
+  clientPreference: string[]
+): string {
+  if (issuerConfig.idTokenSigningAlgValuesSupported === undefined) {
```
Comment (Vinnl) [inline packages/node/src/login/oidc/ClientRegistrar.ts:45]: I'm having a bit trouble navigating the codebase from the PR interview, but have you tested this using an `issuerConfig` in which this property is not declared, rather than undefined? i.e.

```
issuerConfig = {}
```

rather than

```
issuerConfig = { idTokenSigningAlgValuesSupported: undefined }
```

Off the top of my head I'm not sure whether the same holds for object properties as for variables, but if it does, the former might be problematic (which can be solved by doing a check for `typeof ... === "undefined"`).

Comment (Vinnl) [reply packages/node/src/login/oidc/ClientRegistrar.ts:45]: Actually, might be good to use  `Array.isArray()` anyway, just to also guard against some more defined-but-invalid values.

File: packages/node/src/login/oidc/refresh/TokenRefresher.spec.ts [hunk 8]
```diff
@@ -207,6 +213,28 @@ describe("TokenRefresher", () => {
...
+      }),
+    });
+
+    const refreshedTokens = await refresher.refresh(
+      "mySession",
+      "some refresh token",
+      mockJwk()
+    );
+
+    expect(refreshedTokens.access_token).toEqual(mockDpopTokens().access_token);
```
Comment (Vinnl) [inline packages/node/src/login/oidc/refresh/TokenRefresher.spec.ts:235]: Classic :)

```suggestion
    expect(refreshedTokens.access_token).toBe(mockDpopTokens().access_token);
```

File: packages/node/src/login/oidc/refresh/TokenRefresher.ts [hunk 9]
```diff
@@ -80,9 +82,16 @@ export default class TokenRefresher implements ITokenRefresher {
       { sessionId },
       oidcContext.issuerConfig
     );
+    if (clientInfo.idTokenSignedResponseAlg === undefined) {
```
Comment (Vinnl) [inline packages/node/src/login/oidc/refresh/TokenRefresher.ts:85]: Likewise here as the other comment about comparing with `undefined`.

File: packages/oidc/src/common/negotiation.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+// FIXME: this is copy-pasted from @inrupt/solid-client-authn-core (as is all the rest of this directory).
+// Initially, it was aimed that @inrupt/oidc-client-ext would not depend on the core package, and only
+// on the underlying OIDC library, but it is beginning to be impractical. This should be resolved in a
+// future PR.
```
Comment (Vinnl) [inline packages/oidc/src/common/negotiation.ts:25]: Maybe also mention here what the resolution should be in that future PR, i.e. should it be consolidated into -core, or extracted into the individual packages?

**Review by Vinnl — APPROVED** [review]

File: packages/core/src/login/oidc/IClientRegistrar.test.ts (hunk 6 above)
Comment (pmcb55) [inline packages/core/src/login/oidc/IClientRegistrar.test.ts:124]: What if no preference is provided at all...? Needs an explicit test, just to communicate the behaviour (is it 'null' or first preferred entry...?)

File: packages/oidc/src/common/negotiation.ts (hunk 10 above)
Comment (NSeydoux) [reply packages/oidc/src/common/negotiation.ts:25]: Well I wasn't sure what the resolution should be really :). My initial thought was to have `oidc-client-ext` simply depend on `solid-client-authn-core`, but I think there was a reason not to do that earlier, so I preferred not to say anything in the comment, and consolidate this properly later.

File: CHANGELOG.md (hunk 4 above)
Comment (NSeydoux) [reply CHANGELOG.md]: 725b5c335a8a16de7ecfaa20c0235a30a85398b5

File: packages/core/src/login/oidc/IClientRegistrar.ts (hunk 5 above)
Comment (NSeydoux) [reply packages/core/src/login/oidc/IClientRegistrar.ts]: 77abf75ad79c767e18d6515ff269cafdbb260c0e

File: packages/node/src/login/oidc/ClientRegistrar.ts (hunk 7 above)
Comment (NSeydoux) [reply packages/node/src/login/oidc/ClientRegistrar.ts]: a90e36b19f5ac153d425356eb01a746b5f52f1a5

File: packages/node/src/login/oidc/refresh/TokenRefresher.ts (hunk 9 above)
Comment (NSeydoux) [reply packages/node/src/login/oidc/refresh/TokenRefresher.ts:85]: a90e36b19f5ac153d425356eb01a746b5f52f1a5

File: packages/core/src/login/oidc/IClientRegistrar.test.ts (hunk 6 above)
Comment (NSeydoux) [reply packages/core/src/login/oidc/IClientRegistrar.test.ts]: 0cb556ce03d197a6ebb51075ed4de50e0a57d8d1

File: packages/oidc/src/common/negotiation.ts (hunk 10 above)
Comment (Vinnl) [reply packages/oidc/src/common/negotiation.ts:25]: Ah haha, yeah that makes it hard to note down the resolution :P

**Comment by vanhumbeecka** [comment]
I just bumped onto an issue that is probably related to this ticket:

While using `@inrupt/solid-client-authn-node` for server-side authentication, everything works fine when I use the `inrupt.net` pods (probably using RSA?). But fails when using the `broker.pods.inrupt.com` pods (probably using ES?).

My stacktrace shows:
```
11:43:11.376Z ERROR survey-aggregator: Uncaught Error (body={}, host=localhost, method=GET, params={}, path=/handle-redirect)
    RPError: unexpected JWT alg received, expected RS256, got: ES256
        at Client.validateJWT (/Users/andries/projects/consolidate/survey-aggregator/node_modules/openid-client/lib/client.js:814:13)
        at Client.validateIdToken (/Users/andries/projects/consolidate/survey-aggregator/node_modules/openid-client/lib/client.js:693:60)
        at Client.callback (/Users/andries/projects/consolidate/survey-aggregator/node_modules/openid-client/lib/client.js:469:18)
        at processTicksAndRejections (internal/process/task_queues.js:97:5)
        at async AuthCodeRedirectHandler.handle (/Users/andries/projects/consolidate/survey-aggregator/node_modules/@inrupt/solid-client-authn-node/dist/login/oidc/redirectHandler/AuthCodeRedirectHandler.js:79:24)
        at async ClientAuthentication.handleIncomingRedirect (/Users/andries/projects/consolidate/survey-aggregator/node_modules/@inrupt/solid-client-authn-node/dist/ClientAuthentication.js:71:34)
        at async Session.handleIncomingRedirect (/Users/andries/projects/consolidate/survey-aggregator/node_modules/@inrupt/solid-client-authn-node/dist/Session.js:45:35)
        at async AuthController.handleRedirect (/Users/andries/projects/consolidate/survey-aggregator/dist/modules/auth/AuthController.js:50:13)
```

**Comment by NSeydoux** [comment]
Hi @vanhumbeecka, thanks for reporting this ! It looks like this is absolutely related to this issue. The fix should have been released with `v1.7.4` this morning, which version of the library are you depending on ?

**Comment by vanhumbeecka** [comment]
Hi @NSeydoux , I'm using `v1.7.2`. 
I just upgraded to `v1.7.4` to confirm, and it works now :)
